--- vdr-1.7.36/Makefile.orf	2013-01-21 11:56:35.004860899 +0100
+++ vdr-1.7.36/Makefile	2013-01-21 11:57:04.100860032 +0100
@@ -33,10 +33,10 @@
 CONFDIR   ?= /etc/vdr
 CACHEDIR  ?= /var/cache/vdr
 
-PREFIX    ?= /usr/local
+PREFIX    ?= /usr
 BINDIR    ?= $(PREFIX)/bin
 INCDIR    ?= $(PREFIX)/include
-LIBDIR    ?= $(PREFIX)/lib/vdr
+LIBDIR    ?= $(PREFIX)/lib/vdr/plugins
 LOCDIR    ?= $(PREFIX)/share/locale
 MANDIR    ?= $(PREFIX)/share/man
 PCDIR     ?= $(PREFIX)/lib/pkgconfig
