--- media_build-20140208/linux/Makefile.org	2014-05-15 17:16:46.178402998 +0200
+++ media_build-20140208/linux/Makefile	2014-05-15 17:17:11.866403444 +0200
@@ -57,8 +57,8 @@
 
 TODAY_TAR := linux-media-`date "+%Y-%m-%d"`.tar
 TODAY_TAR_MD5 := linux-media-`date "+%Y-%m-%d"`.tar.bz2.md5
-LATEST_TAR := http://linuxtv.org/downloads/drivers/linux-media-LATEST.tar.bz2
-LATEST_TAR_MD5 := http://linuxtv.org/downloads/drivers/linux-media-LATEST.tar.bz2.md5
+LATEST_TAR := http://linuxtv.org/downloads/drivers/old/linux-media-LATEST.tar.bz2
+LATEST_TAR_MD5 := http://linuxtv.org/downloads/drivers/old/linux-media-LATEST.tar.bz2.md5
 
 default:
 	make -C ..
