diff -ruN vdr/config.c vdr-1/config.c
--- vdr/config.c	2014-01-08 21:49:36.000000000 +0100
+++ vdr-1/config.c	2014-01-08 21:51:00.000000000 +0100
@@ -398,6 +398,7 @@
   PositionerLastLon = 0;
   SetSystemTime = 0;
   TimeSource = 0;
+  ChannelBlocker = 0;
   TimeTransponder = 0;
   StandardCompliance = STANDARD_DVB;
   MarginStart = 2;
@@ -618,6 +619,7 @@
   else if (!strcasecmp(Name, "SetSystemTime"))       SetSystemTime      = atoi(Value);
   else if (!strcasecmp(Name, "TimeSource"))          TimeSource         = cSource::FromString(Value);
   else if (!strcasecmp(Name, "TimeTransponder"))     TimeTransponder    = atoi(Value);
+  else if (!strcasecmp(Name, "ChannelBlocker"))      ChannelBlocker     = atoi(Value);
   else if (!strcasecmp(Name, "StandardCompliance"))  StandardCompliance = atoi(Value);
   else if (!strcasecmp(Name, "MarginStart"))         MarginStart        = atoi(Value);
   else if (!strcasecmp(Name, "MarginStop"))          MarginStop         = atoi(Value);
@@ -741,6 +743,7 @@
   Store("SetSystemTime",      SetSystemTime);
   Store("TimeSource",         cSource::ToString(TimeSource));
   Store("TimeTransponder",    TimeTransponder);
+  Store("ChannelBlocker",     ChannelBlocker);
   Store("StandardCompliance", StandardCompliance);
   Store("MarginStart",        MarginStart);
   Store("MarginStop",         MarginStop);
diff -ruN vdr/config.h vdr-1/config.h
--- vdr/config.h	2014-01-08 21:49:36.000000000 +0100
+++ vdr-1/config.h	2014-01-08 21:51:00.000000000 +0100
@@ -284,6 +284,7 @@
   int SetSystemTime;
   int TimeSource;
   int TimeTransponder;
+  int ChannelBlocker;
   int StandardCompliance;
   int MarginStart, MarginStop;
   int JumpSeconds, JumpSecondsSlow, JumpSecondsRepeat;
diff -ruN vdr/dvbdevice.c vdr-1/dvbdevice.c
--- vdr/dvbdevice.c	2014-01-02 11:30:15.000000000 +0100
+++ vdr-1/dvbdevice.c	2014-01-08 21:51:00.000000000 +0100
@@ -1510,6 +1510,10 @@
 
 bool cDvbDevice::ProvidesSource(int Source) const
 {
+  if (Setup.ChannelBlocker == 1) {
+      if (IsPrimaryDevice()) return false;
+      //isyslog("ChannelBlocker active on primary Interface");
+  }
   int type = Source & cSource::st_Mask;
   return type == cSource::stNone
       || type == cSource::stAtsc  && ProvidesDeliverySystem(SYS_ATSC)
diff -ruN vdr/menu.c vdr-1/menu.c
--- vdr/menu.c	2014-01-08 21:49:36.000000000 +0100
+++ vdr-1/menu.c	2014-01-08 21:51:00.000000000 +0100
@@ -3290,6 +3290,7 @@
      Add(new cMenuEditIntItem( tr("Setup.DVB$Subtitle background transparency"), &data.SubtitleBgTransparency, 0, 10));
      }
   Add(new cMenuEditBoolItem(tr("Setup.DVB$Enable teletext support"), &data.SupportTeletext));
+  Add(new cMenuEditBoolItem(tr("Setup.DVB$Channel Blocker"), &data.ChannelBlocker));
 
   SetCurrent(Get(current));
   Display();
diff -ruN vdr/po/de_DE.po vdr-1/po/de_DE.po
--- vdr/po/de_DE.po	2014-01-08 21:49:36.000000000 +0100
+++ vdr-1/po/de_DE.po	2014-01-08 21:51:00.000000000 +0100
@@ -966,6 +966,9 @@
 msgid "Setup.EPG$Preferred language"
 msgstr "Bevorzugte Sprache"
 
+msgid "Setup.DVB$Channel Blocker"
+msgstr "Tuner des primären DVB blocken"
+
 msgid "pan&scan"
 msgstr "pan&scan"
 
