--- xbmc-pvr-addons-git+04258ed894/addons/pvr.vdr.vnsi/vdr-plugin-vnsiserver/Makefile.org	2013-01-04 00:21:48.746600212 +0100
+++ xbmc-pvr-addons-git+04258ed894/addons/pvr.vdr.vnsi/vdr-plugin-vnsiserver/Makefile	2013-01-04 00:22:20.626599262 +0100
@@ -36,7 +36,7 @@
 endif
 
 # backwards compatibility version < 1.7.34
-API1733 := $(shell if [ "$(APIVERSION)" \< "1.7.34" ]; then echo true; fi; )
+API1733 := $(shell if [ "$(APIVERSION)" \< "200" ]; then echo true; fi; )
 
 ifdef API1733
 
