--- vdr-plugin-xvdr-0.9.8+reba050c847/Makefile	2013-01-29 20:11:27.000000000 +0100
+++ vdr-plugin-xvdr-0.9.8+reba050c847/Makefile.org	2013-01-29 23:27:42.476916971 +0100
@@ -23,6 +23,8 @@
 #
 TMPDIR ?= /tmp
 
+include $(VDRDIR)/Make.config
+
 ### The compiler options:
 
 export CFLAGS   = $(call PKGCFG,cflags)
@@ -38,22 +40,25 @@
 
 ### The name of the distribution archive:
 
-ARCHIVE = $(PLUGIN)-$(VERSION)
+ARCHIVE = $(APIVERSION)
 PACKAGE = vdr-$(ARCHIVE)
 
 ### The name of the shared object file:
 
-SOFILE = libvdr-$(PLUGIN).so
+SOFILE = libvdr-$(PLUGIN).so.$(ARCHIVE)
 
 ### Includes and Defines (add further entries here):
 
 INCLUDES += -I./src
+INCLUDES += -I$(VDRDIR)/include -I$(DVBDIR)/include -I$(VDRDIR) -I./src -I.
 
 ifdef DEBUG
 INCLUDES += -DDEBUG
 endif
 
-DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"' -DXVDR_VERSION='"$(VERSION)"'
+#DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"' -DXVDR_VERSION='"$(VERSION)"'
+DEFINES += -DPLUGIN_NAME_I18N='"$(PLUGIN)"' -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DXVDR_VERSION='"$(VERSION)"'
+
 
 ### The object files (add further files here):
 
