--- cedarx-libs-3.0.43/libcedarv/linux-armhf/Makefile.org	2013-05-29 12:12:53.458554536 +0200
+++ cedarx-libs-3.0.43/libcedarv/linux-armhf/Makefile	2013-05-29 12:13:01.702554290 +0200
@@ -1,9 +1,9 @@
 #
 # Assume standlone toolchain
 #
-CC  = $(CROSS_COMPILE)gcc
-CXX = $(CROSS_COMPILE)g++
-AR  = $(CROSS_COMPILE)ar
+#CC  = $(CROSS_COMPILE)gcc
+#CXX = $(CROSS_COMPILE)g++
+#AR  = $(CROSS_COMPILE)ar
 
 CEDARDIR=$(shell pwd)
 
