--- linux-3.6.11/drivers/media/common/tuners/mxl5007t.c.org	2013-01-14 14:58:26.409507728 +0100
+++ linux-3.6.11/drivers/media/common/tuners/mxl5007t.c	2013-01-14 15:00:44.021510094 +0100
@@ -374,7 +374,7 @@
 	mxl5007t_set_if_freq_bits(state, cfg->if_freq_hz, cfg->invert_if);
 	mxl5007t_set_xtal_freq_bits(state, cfg->xtal_freq_hz);
 
-	set_reg_bits(state->tab_init, 0x04, 0x01, cfg->loop_thru_enable);
+	//set_reg_bits(state->tab_init, 0x04, 0x01, cfg->loop_thru_enable);
 	set_reg_bits(state->tab_init, 0x03, 0x08, cfg->clk_out_enable << 3);
 	set_reg_bits(state->tab_init, 0x03, 0x07, cfg->clk_out_amp);
 
@@ -531,9 +531,13 @@
 	struct reg_pair_t *init_regs;
 	int ret;
 
-	ret = mxl5007t_soft_reset(state);
-	if (mxl_fail(ret))
+	//ret = mxl5007t_soft_reset(state);
+	//if (mxl_fail(ret))
+        if (!state->config->no_reset) {
+                ret = mxl5007t_soft_reset(state);
+                if (mxl_fail(ret))
 		goto fail;
+        }
 
 	/* calculate initialization reg array */
 	init_regs = mxl5007t_calc_init_regs(state, mode);
@@ -887,7 +891,13 @@
 		if (fe->ops.i2c_gate_ctrl)
 			fe->ops.i2c_gate_ctrl(fe, 1);
 
-		ret = mxl5007t_get_chip_id(state);
+		//ret = mxl5007t_get_chip_id(state);
+                if (!state->config->no_probe)
+                        ret = mxl5007t_get_chip_id(state);
+
+                ret = mxl5007t_write_reg(state, 0x04,
+                        state->config->loop_thru_enable);
+
 
 		if (fe->ops.i2c_gate_ctrl)
 			fe->ops.i2c_gate_ctrl(fe, 0);
--- linux-3.6.11/drivers/media/common/tuners/mxl5007.h.org	2013-01-14 15:01:01.729510397 +0100
+++ linux-3.6.11/drivers/media/common/tuners/mxl5007t.h	2013-01-14 15:01:56.017511327 +0100
@@ -73,8 +73,11 @@
 	enum mxl5007t_xtal_freq xtal_freq_hz;
 	enum mxl5007t_if_freq if_freq_hz;
 	unsigned int invert_if:1;
-	unsigned int loop_thru_enable:1;
+	//unsigned int loop_thru_enable:1;
+        unsigned int loop_thru_enable:3;
 	unsigned int clk_out_enable:1;
+        unsigned int no_probe:1;
+        unsigned int no_reset:1;
 };
 
 #if defined(CONFIG_MEDIA_TUNER_MXL5007T) || (defined(CONFIG_MEDIA_TUNER_MXL5007T_MODULE) && defined(MODULE))
