--- enigma2pc-20130727/scripts/dvb_my.sh.org	2013-10-24 14:40:48.191502731 +0200
+++ enigma2pc-20130727/scripts/dvb_my.sh	2013-10-24 14:41:59.011503947 +0200
@@ -1,23 +1,23 @@
 #!/bin/sh
 
 if [ $(lsmod | grep -c dvbsoftwareca) -eq 0 ]; then
-	sudo modprobe dvbsoftwareca
+	modprobe dvbsoftwareca
 fi
 
 for i in $(find /dev/dvb -name demux0 | sed 's/.\{6\}$//')
 do
         echo "For $i"demux0" create symlink $i"demux1" "
-        sudo ln -s $i"demux0" $i"demux1"
+        ln -s $i"demux0" $i"demux1"
         echo "For $i"dvr0" create symlink $i"dvr1" "
-        sudo ln -s $i"dvr0" $i"dvr1"
+        ln -s $i"dvr0" $i"dvr1"
 done
 
 sleep 1
 
 if [ $(ps -A | grep -c oscam) -ne 0 ]; then
-	sudo /etc/init.d/softcam.oscam restart &
+	/bin/systemctl restart oscam.service
 else
-	sudo /etc/init.d/softcam.oscam start &
+	/bin/systemctl start oscam.service
 fi
 
 exit 0
