--- linux/drivers/gpu/drm/vivante/vivante_drv.c.orig	2014-01-26 13:43:27.758873772 -0700
+++ linux/drivers/gpu/drm/vivante/vivante_drv.c	2014-01-26 13:43:44.738874193 -0700
@@ -55,7 +55,7 @@
 
 #include "drm_pciids.h"
 
-static char platformdevicename[] = "Vivante GCCore";
+static char platformdevicename[] = "Vivante GCCore:00";
 static struct platform_device *pplatformdev;
 
 static struct drm_driver driver = {
