--- tvheadend-3.6+7d8d81a/support/configure.inc.org	2014-02-27 11:37:58.322359133 +0100
+++ tvheadend-3.6+7d8d81a/support/configure.inc	2014-02-27 11:38:07.682358854 +0100
@@ -265,7 +265,7 @@
 EOF
   $CC $CFLAGS $LDFLAGS $TMPDIR/$$.c -o $TMPDIR/$$.bin $opt &> /dev/null
   RET=$?
-  [ $RET -eq 0 ] && $TMPDIR/$$.bin
+  #[ $RET -eq 0 ] && $TMPDIR/$$.bin
   RET=$?
   rm -f $TMPDIR/$$.{c,bin}
   return $RET
