--- linux-3.7.6/drivers/media/usb/dvb-usb-v2/rtl28xxu.c.orig	2013-02-19 12:09:35.000000000 +0100
+++ linux-3.7.6/drivers/media/usb/dvb-usb-v2/rtl28xxu.c	2013-02-19 17:38:13.232599298 +0100
@@ -1364,6 +1364,8 @@
 		&rtl2832u_props, "Dexatek DK mini DVB-T Dongle", NULL) },
 	{ DVB_USB_DEVICE(USB_VID_TERRATEC, 0x00d7,
 		&rtl2832u_props, "TerraTec Cinergy T Stick+", NULL) },
+	{ DVB_USB_DEVICE(USB_VID_KWORLD_2, 0xd394,
+		&rtl2832u_props, "MaxMedia HU394-T", NULL) },
 	{ DVB_USB_DEVICE(USB_VID_KWORLD_2, 0xd3a8,
 		&rtl2832u_props, "ASUS My Cinema-U3100Mini Plus V2", NULL) },
 	{ DVB_USB_DEVICE(USB_VID_KWORLD_2, 0xd393,
