--- vdr-dvbapi-1.0.4/Makefile.org	2013-02-17 16:05:13.000000000 +0100
+++ vdr-dvbapi-1.0.4/Makefile	2013-05-27 10:50:42.307851337 +0200
@@ -85,8 +85,8 @@
 
 ifndef LIBDVBCSA
 # FFdeCSA
-PARALLEL   ?= PARALLEL_128_SSE2
-CSAFLAGS   ?= -fexpensive-optimizations -funroll-loops -mmmx -msse -msse2 -msse3
+PARALLEL   ?= PARALLEL_32_INT
+CSAFLAGS   ?= -fexpensive-optimizations -funroll-loops
 FFDECSADIR  = FFdecsa
 FFDECSA     = $(FFDECSADIR)/FFdecsa.o
 DECSALIB    = $(FFDECSA)
